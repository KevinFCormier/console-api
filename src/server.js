/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import https from 'https';
import fs from 'fs';
import schema from './schema/index';
import config from '../config';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const GRAPHQL_PORT = process.env.PORT || config.get('httpPort') || 4000;
const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(req => ({ formatError, schema, context: req })));

const privateKey = fs.readFileSync(process.env.serverKey || './sslcert/hcmuiapi.key', 'utf8');
const certificate = fs.readFileSync(process.env.serverCert || './sslcert/hcmuiapi.crt', 'utf8');
const credentials = { key: privateKey, cert: certificate };
const server = https.createServer(credentials, graphQLServer);

server.listen(GRAPHQL_PORT, () => {
  logger.info(`[pid ${process.pid}] [env ${process.env.NODE_ENV}] started.`);
  logger.info(`HCM UI API is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphql`);
  if (process.env.NODE_ENV !== 'production') {
    logger.info(`GraphiQL is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphiql`);
  }
});
