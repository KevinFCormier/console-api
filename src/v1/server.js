/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import schema from './schema/index';
import config from '../../config';
import HCMConnector from './datasource/lib/hcm-connector';
import MongoConnector from './datasource/lib/mongo-connector';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};


const getProductionOptions = async (req) => {
  const mongoConnector = new MongoConnector();
  const hcmConnector = new HCMConnector({
    hcmUrl: config.get('hcmUrl') || 'http://localhost:8080',
    pollInterval: config.get('hcmPollInterval') || 200,
    pollTimeout: config.get('hcmPollTimeout') || 10000,
  });

  await mongoConnector.connect(config.get('mongodbUrl') || 'mongodb://localhost:27017/weave');
  return {
    formatError,
    schema,
    context: {
      req,
      hcmConnector,
      mongoConnector,
    },
  };
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(getProductionOptions));

export default graphQLServer;
