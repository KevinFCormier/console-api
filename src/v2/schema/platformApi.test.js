/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Platform Api Resolver', () => {
  test('Get Automated Import Status', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            getAutomatedImportStatus(namespace: "test", name:"test")
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Get Automated Import Status should return error msg', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            getAutomatedImportStatus(namespace: "", name:"test")
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Create Kubernetes Cluster Resource for Import', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
            createClusterResource(body: "test")
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Create Kubernetes Cluster Resource for Import should return error msg', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
            createClusterResource(body: "")
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Start Automated Import Process', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
            automatedImport(namespace:"test", name:"test", body: "test")
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Start Automated Import Process should return error msg', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
            automatedImport(namespace:"", name:"test", body: "test")
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});
