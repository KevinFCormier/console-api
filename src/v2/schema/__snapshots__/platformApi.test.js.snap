// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Platform Api Resolver Correctly Resolves Create Cloud Connection 1`] = `
Object {
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 121,
          "line": 3,
        },
      ],
      "message": "Syntax Error: Unterminated string.",
    },
  ],
}
`;

exports[`Platform Api Resolver Correctly Resolves Delete Cloud Connection 1`] = `
Object {
  "data": Object {
    "deleteCloudConnection": null,
  },
}
`;

exports[`Platform Api Resolver Correctly Resolves Edit Cloud Connection 1`] = `
Object {
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 119,
          "line": 3,
        },
      ],
      "message": "Syntax Error: Unterminated string.",
    },
  ],
}
`;

exports[`Platform Api Resolver Correctly Resolves Get Cloud Connections 1`] = `
Object {
  "data": Object {
    "connections": Array [
      Object {
        "metadata": Object {
          "name": null,
          "name_namespace": null,
          "namespace": null,
          "provider": null,
        },
      },
    ],
  },
}
`;

exports[`Platform Api Resolver Correctly Resolves Get Cloud Providers 1`] = `
Object {
  "data": Object {
    "providers": Array [
      Object {
        "clusterMetadata": "",
        "clusterValues": "",
        "configMetadata": "",
        "configValues": "",
        "longname": "iks-test",
        "name": "iks-test",
        "statusCode": 200,
      },
    ],
  },
}
`;

exports[`Platform Api Resolver Create Kubernetes Cluster Resource for Import 1`] = `
Object {
  "data": Object {
    "createClusterResource": Object {
      "error": Object {
        "rawResponse": Object {
          "apiVersion": "clusterregistry.k8s.io/v1alpha1",
          "items": Array [
            Object {
              "apiVersion": "clusterregistry.k8s.io/v1alpha1",
              "kind": "Cluster",
              "metadata": Object {
                "annotations": Object {
                  "mcm.ibm.com/user-group": "c3lzdGVtOnNlcnZpY2VhY2NvdW50cyxzeXN0ZW06c2VydmljZWFjY291bnRzOmt1YmUtc3lzdGVtLHN5c3RlbTphdXRoZW50aWNhdGVk",
                  "mcm.ibm.com/user-identity": "c3lzdGVtOnNlcnZpY2VhY2NvdW50Omt1YmUtc3lzdGVtOmRlZmF1bHQ=",
                },
                "creationTimestamp": "2018-08-13T16:56:41Z",
                "finalizers": Array [
                  "finalizer.hcm.ibm.com",
                ],
                "labels": Object {
                  "cloud": "IBM",
                  "datacenter": "toronto",
                  "environment": "Dev",
                  "name": "mycluster.icp",
                  "owner": "marketing",
                  "region": "US",
                  "vendor": "ICP",
                },
                "name": "mycluster.icp",
                "namespace": "default",
                "resourceVersion": "136558",
                "selfLink": "/apis/clusterregistry.k8s.io/v1alpha1/namespaces/default/clusters/mycluster.icp",
                "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
              },
              "spec": Object {
                "authInfo": Object {},
                "kubernetesApiEndpoints": Object {
                  "serverEndpoints": Array [
                    Object {
                      "serverAddress": "9.42.80.212:8001",
                    },
                  ],
                },
              },
              "status": Object {
                "conditions": Array [
                  Object {
                    "lastHeartbeatTime": "2018-08-15T19:41:20Z",
                    "lastTransitionTime": null,
                    "status": "",
                    "type": "OK",
                  },
                ],
              },
            },
          ],
          "kind": "ClusterList",
          "metadata": Object {
            "resourceVersion": "136667",
            "selfLink": "/apis/clusterregistry.k8s.io/v1alpha1/clusters",
          },
          "statusCode": "",
          "statusMessage": "",
        },
        "statusCode": "",
      },
    },
  },
}
`;

exports[`Platform Api Resolver Create Kubernetes Cluster Resource for Import should return error msg 1`] = `
Object {
  "data": Object {
    "createClusterResource": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Body is required for createClusterResource",
      "path": Array [
        "createClusterResource",
      ],
    },
  ],
}
`;

exports[`Platform Api Resolver Get Automated Import Status 1`] = `
Object {
  "data": Object {
    "getAutomatedImportStatus": Object {
      "error": Object {
        "rawResponse": Object {
          "apiVersion": "clusterregistry.k8s.io/v1alpha1",
          "items": Array [
            Object {
              "apiVersion": "clusterregistry.k8s.io/v1alpha1",
              "kind": "Cluster",
              "metadata": Object {
                "annotations": Object {
                  "mcm.ibm.com/user-group": "c3lzdGVtOnNlcnZpY2VhY2NvdW50cyxzeXN0ZW06c2VydmljZWFjY291bnRzOmt1YmUtc3lzdGVtLHN5c3RlbTphdXRoZW50aWNhdGVk",
                  "mcm.ibm.com/user-identity": "c3lzdGVtOnNlcnZpY2VhY2NvdW50Omt1YmUtc3lzdGVtOmRlZmF1bHQ=",
                },
                "creationTimestamp": "2018-08-13T16:56:41Z",
                "finalizers": Array [
                  "finalizer.hcm.ibm.com",
                ],
                "labels": Object {
                  "cloud": "IBM",
                  "datacenter": "toronto",
                  "environment": "Dev",
                  "name": "mycluster.icp",
                  "owner": "marketing",
                  "region": "US",
                  "vendor": "ICP",
                },
                "name": "mycluster.icp",
                "namespace": "default",
                "resourceVersion": "136558",
                "selfLink": "/apis/clusterregistry.k8s.io/v1alpha1/namespaces/default/clusters/mycluster.icp",
                "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
              },
              "spec": Object {
                "authInfo": Object {},
                "kubernetesApiEndpoints": Object {
                  "serverEndpoints": Array [
                    Object {
                      "serverAddress": "9.42.80.212:8001",
                    },
                  ],
                },
              },
              "status": Object {
                "conditions": Array [
                  Object {
                    "lastHeartbeatTime": "2018-08-15T19:41:20Z",
                    "lastTransitionTime": null,
                    "status": "",
                    "type": "OK",
                  },
                ],
              },
            },
          ],
          "kind": "ClusterList",
          "metadata": Object {
            "resourceVersion": "136667",
            "selfLink": "/apis/clusterregistry.k8s.io/v1alpha1/clusters",
          },
          "statusCode": "",
          "statusMessage": "",
        },
        "statusCode": "",
      },
    },
  },
}
`;

exports[`Platform Api Resolver Get Automated Import Status should return error msg 1`] = `
Object {
  "data": Object {
    "getAutomatedImportStatus": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "namespace and name are required for getAutomatedImportStatus",
      "path": Array [
        "getAutomatedImportStatus",
      ],
    },
  ],
}
`;

exports[`Platform Api Resolver Start Automated Import Process 1`] = `
Object {
  "data": Object {
    "automatedImport": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Cannot read property 'includes' of undefined",
      "path": Array [
        "automatedImport",
      ],
    },
  ],
}
`;

exports[`Platform Api Resolver Start Automated Import Process should return error msg 1`] = `
Object {
  "data": Object {
    "automatedImport": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 13,
          "line": 3,
        },
      ],
      "message": "Body, namespace, and name are required for automatedImport",
      "path": Array [
        "automatedImport",
      ],
    },
  ],
}
`;

exports[`Platform Api Resolver createCluster test that we can create cluster correctly 1`] = `
Object {
  "data": Object {
    "createCluster": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 17,
          "line": 3,
        },
      ],
      "message": "Cannot read property 'includes' of undefined",
      "path": Array [
        "createCluster",
      ],
    },
  ],
}
`;

exports[`Platform Api Resolver createCluster test that we get an error when cluster not provided 1`] = `
Object {
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 54,
          "line": 3,
        },
      ],
      "message": "Expected type JSON!, found null.",
    },
  ],
}
`;

exports[`Platform Api Resolver createCluster test that we get an error when namespace not provided 1`] = `
Object {
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 45,
          "line": 3,
        },
      ],
      "message": "Syntax Error: Expected Name, found {",
    },
  ],
}
`;

exports[`Platform Api Resolver createCluster test with all required fields but returns with error 1`] = `
Object {
  "data": Object {
    "createCluster": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 17,
          "line": 3,
        },
      ],
      "message": "Cannot read property 'includes' of undefined",
      "path": Array [
        "createCluster",
      ],
    },
  ],
}
`;
